eventsources/sources/sftp/start.go

```diff
@@ -0,0 +1,282 @@
+/*
+Copyright 2018 BlackRock, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sftp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/sftp"
+	"go.uber.org/zap"
+	"golang.org/x/crypto/ssh"
+
+	"github.com/argoproj/argo-events/common"
+	"github.com/argoproj/argo-events/common/logging"
+	eventsourcecommon "github.com/argoproj/argo-events/eventsources/common"
+	"github.com/argoproj/argo-events/eventsources/common/fsevent"
+	"github.com/argoproj/argo-events/eventsources/sources"
+	metrics "github.com/argoproj/argo-events/metrics"
+	apicommon "github.com/argoproj/argo-events/pkg/apis/common"
+	"github.com/argoproj/argo-events/pkg/apis/eventsource/v1alpha1"
+)
+
+// EventListener implements Eventing for sftp event source
+type EventListener struct {
+	EventSourceName string
+	EventName       string
+	SFTPEventSource v1alpha1.SFTPEventSource
+	Metrics         *metrics.Metrics
+}
+
+// GetEventSourceName returns name of event source
+func (el *EventListener) GetEventSourceName() string {
+	return el.EventSourceName
+}
+
+// GetEventName returns name of event
+func (el *EventListener) GetEventName() string {
+	return el.EventName
+}
+
+// GetEventSourceType return type of event server
+func (el *EventListener) GetEventSourceType() apicommon.EventSourceType {
+	return apicommon.SFTPEvent
+}
+
+// StartListening starts listening events
+func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byte, ...eventsourcecommon.Option) error) error {
+	log := logging.FromContext(ctx).
+		With(logging.LabelEventSourceType, el.GetEventSourceType(), logging.LabelEventName, el.GetEventName())
+	defer sources.Recover(el.GetEventName())
+
+	username, err := common.GetSecretFromVolume(el.SFTPEventSource.Username)
+	if err != nil {
+		return fmt.Errorf("username not found, %w", err)
+	}
+	password, err := common.GetSecretFromVolume(el.SFTPEventSource.Password)
+	if err != nil {
+		return fmt.Errorf("password not found, %w", err)
+	}
+	address, err := common.GetSecretFromVolume(el.SFTPEventSource.Address)
+	if err != nil {
+		return fmt.Errorf("address not found, %w", err)
+	}
+
+	sftpConfig := &ssh.ClientConfig{
+		User:            username,
+		Auth:            []ssh.AuthMethod{ssh.Password(password)},
```

Comment:
would be good to have option for sshkey based auth instead of password based too: https://pkg.go.dev/golang.org/x/crypto/ssh#PublicKeys

Reply:
https://github.com/argoproj/argo-events/pull/2918